homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -47,36 +50,76 @@ async def async_step_user(self, user_input=None):
             step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
         )
 
+    async def async_step_zeroconf(self, user_input=None):
+        """Handle zeroconf discovery."""
+        if user_input is None:
+            return self.async_abort(reason="not_xiaomi_miio")
+
+        name = user_input.get("name")
+        self.host = user_input.get("host")
+        mac_address = user_input.get("properties", {}).get("mac")
+
+        if not name or not self.host or not mac_address:
+            return self.async_abort(reason="not_xiaomi_miio")
+
+        # format mac (include semicolns and make uppercase)
+        if len(mac_address) == 12:
+            mac_address = ":".join(mac_address[i : i + 2] for i in range(0, 12, 2))
+        mac_address = mac_address.upper()
+
+        # Check which device is discovered.
+        if name.startswith(ZEROCONF_GATEWAY):
+            unique_id = f"{mac_address}"
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured()
```

Comment:
Update the existing entry if you find one, so you keep having latest host info.

```suggestion
            self._abort_if_unique_id_configured({CONF_HOST: host})
```